app/src/main/java/org/simple/clinic/protocolv2/ProtocolRepository.kt

```diff
@@ -0,0 +1,64 @@
+package org.simple.clinic.protocolv2
+
+import io.reactivex.Completable
+import io.reactivex.Observable
+import io.reactivex.Single
+import org.simple.clinic.patient.SyncStatus
+import org.simple.clinic.patient.canBeOverriddenByServerCopy
+import org.simple.clinic.protocolv2.sync.ProtocolPayload
+import org.simple.clinic.sync.SynceableRepository
+import java.util.UUID
+import javax.inject.Inject
+
+class ProtocolRepository @Inject constructor(
+    private val protocolDao: Protocol.RoomDao,
+    private val protocolDrugsDao: ProtocolDrug.RoomDao
+) : SynceableRepository<ProtocolDrugsWithDosage, ProtocolPayload> {
+
+  override fun save(records: List<ProtocolDrugsWithDosage>): Completable {
+    return Completable.fromAction {
+      protocolDao.save(records
+          .map { it.protocol })
+      protocolDrugsDao.save(records
+          .filter { it.drugs.isNotEmpty() }
+          .flatMap { it.drugs })
+    }
+  }
+
+  override fun recordsWithSyncStatus(syncStatus: SyncStatus): Single<List<ProtocolDrugsWithDosage>> {
+    return Single.fromCallable {
+      emptyList<ProtocolDrugsWithDosage>()
+    }
+  }
+
+  override fun setSyncStatus(from: SyncStatus, to: SyncStatus): Completable {
+    return Completable.fromAction { protocolDao.updateSyncStatus(oldStatus = from, newStatus = to) }
+  }
+
+  override fun setSyncStatus(ids: List<UUID>, to: SyncStatus): Completable {
+    return Completable.fromAction { protocolDao.updateSyncStatus(uuids = ids, newStatus = to) }
+  }
+
+  override fun mergeWithLocalData(payloads: List<ProtocolPayload>): Completable {
+    return Single.fromCallable {
+      payloads
+          .filter { payload ->
+            protocolDao
+                .getOne(payload.uuid)?.syncStatus.canBeOverriddenByServerCopy()
+          }
+          .map(::payloadToProtocolDrugWithDosage)
```

Comment:
Creating a Single just to emit a piece of data is not worth it. I assume this was being done in the place where you copied this from. You could pass the object directly to `save()`.

```kotlin
val newOrUpdatedProtocols = payloads
    .filter { payload ->
      protocolDao
          .getOne(payload.uuid)?.syncStatus.canBeOverriddenByServerCopy()
    }
    .map(::payloadToProtocolDrugWithDosage)
return save(newOrUpdatedProtocols)
```

Reply:
I would also break this line: 

```kotlin
.filter { payload -> 
    protocolDao
          .getOne(payload.uuid)?.syncStatus.canBeOverriddenByServerCopy()
}
```

(Side note: we move operators to the next line only if we have >= 2 operators)

into:

```kotlin
.filter { payload ->
  val localCopy = protocolDao.getOne(payload.uuid)
  localCopy?.syncStatus.canBeOverriddenByServerCopy()
}
```